### PR TITLE
Simplify Message.body_length

### PR DESCRIPTION
**Pull request by reynir**
Instead of doing sort_uniq on all content-length headers we can instead check that they are all equal (in linear time \o/)

### CODE REVIEW

File: lib/message.ml [hunk 1]
```diff
@@ -36,22 +36,19 @@
...
 
 let body_length headers =
-  match unique_content_length_values headers with
-  | [ len ] ->
-    let len = content_length_of_string len in
-    if Int64.compare len 0L >= 0 then `Fixed len else `Error `Bad_request
-  | _ -> `Unknown
+  match Headers.get_multi headers "content-length" with
+  | [] -> `Unknown
+  | [ x ] ->  content_length_of_string x
```
Comment (reynir) [inline lib/message.ml:47]: This branch could be removed.

File: lib/message.ml [hunk 2]
```diff
@@ -36,22 +36,19 @@
...
-  match xs with [] | [ _ ] -> xs | _ -> List.sort_uniq String.compare xs
-
-let unique_content_length_values headers =
-  (* XXX(seliopou): perform proper content-length parsing *)
-  sort_uniq (Headers.get_multi headers "content-length")
-
-let content_length_of_string s = try Int64.of_string s with _ -> -1L
+let content_length_of_string s =
+  match Int64.of_string s with
+  | len when len >= 0L -> `Fixed len
```
Comment (anmonteiro) [inline lib/message.ml:41]: can we avoid the polymorphic comparison here? the previous version of this code used `Int64.compare`

**Comment by anmonteiro** [comment]
I think I'm missing something. How is this change equivalent to the previous code if the content-length values aren't sent in the same order?

**Comment by anmonteiro** [comment]
I was indeed missing something. From [RFC 7230§3.3.2](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.2)

>  If a message is received that has multiple Content-Length header
   fields with field-values consisting of the same decimal value, or a
   single Content-Length header field with a field value containing a
   list of identical decimal values (e.g., "Content-Length: 42, 42"),
   indicating that duplicate Content-Length header fields have been
   generated or combined by an upstream message processor, then the
   recipient MUST either reject the message as invalid or replace the
   duplicated field-values with a single valid Content-Length field
   containing that decimal value prior to determining the message body
   length or forwarding the message.

So one can only really have a single value for content-length, though it might be repeated. In which case your PR makes sense.

File: lib/message.ml (hunk 1 above)
Comment (anmonteiro) [reply lib/message.ml:47]: happy to keep it around. I feel like it improves readability.

**Comment by reynir** [comment]
Ah, you are right. I thought this was equivalent to the previous version, but indeed this is more strict and returns an error where the previous version returned `` `Unknown``.

**Comment by anmonteiro** [comment]
> Ah, you are right. I thought this was equivalent to the previous version, but indeed this is more strict and returns an error where the previous version returned `` `Unknown ``.

I think we can restore the previous behavior by returning `` `Unknown `` in the `else` branch, right?

File: lib/message.ml [hunk 3]
```diff
@@ -36,22 +36,19 @@
...
+  match Headers.get_multi headers "content-length" with
+  | [] -> `Unknown
+  | [ x ] ->  content_length_of_string x
+  | hd :: tl ->
+    (* if there are multiple content-length headers we require them all to be
+     * exactly equal. *)
+    if List.for_all (String.equal hd) tl then
+      content_length_of_string hd
+    else
+      `Error `Bad_request
```
Comment (reynir) [inline lib/message.ml:54]: Yes, we can restore the previous behavior like so.
```suggestion
      `Uknown
```

**Comment by anmonteiro** [comment]
Thanks!

**Pull request merged**